File: Sources/SomnioServerCore/Connections/AITickService.swift
import Foundation
import Logging
import ServiceLifecycle

/// Periodic AI-tick driver. Each pass calls `WorldRouter.runAITickAcrossSectors()`, which
/// iterates every loaded sector's actor and persists the per-tick dialog digests. The
/// shutdown-cancel + error semantics live in `runPeriodically`.
public actor AITickService: Service {
    /// The AI-tick cadence in seconds — a server-loop implementation detail, not shared
    /// gameplay policy. Paired with `SomnioConstants.npcDialogCooldownSeconds` so the dialog
    /// cooldown stays a single rule: `PerSectorActor.NPCRuntime.dialogCooldownCap` derives its
    /// tick count from this value. `public` so the `public init` default below can reference it.
    public static let defaultAITickIntervalSeconds = 0.05

    private let worldRouter: WorldRouter
    private let interval: Duration
    private let logger: Logger

    public init(
        worldRouter: WorldRouter,
        interval: Duration = .seconds(AITickService.defaultAITickIntervalSeconds),
        logger: Logger
    ) {
        self.worldRouter = worldRouter
        self.interval = interval
        self.logger = logger
    }

    public func run() async throws {
        await runPeriodically(interval: interval, logger: logger, label: "ai tick") { [worldRouter] in
            await worldRouter.runAITickAcrossSectors()
        }
    }
}

File: packages/server/src/services/aiTickService.ts
import type { WorldRouter } from '../world/worldRouter.ts';
import { DEFAULT_AI_TICK_INTERVAL_SECONDS } from '../world/perSectorActor.ts';
import { runPeriodically } from './periodicLoop.ts';

/** Drives `runAITickAcrossSectors` on the 50 ms cadence until aborted. */
export class AITickService {
  private readonly worldRouter: WorldRouter;
  private readonly intervalMs: number;

  constructor(worldRouter: WorldRouter, intervalMs: number = DEFAULT_AI_TICK_INTERVAL_SECONDS * 1000) {
    this.worldRouter = worldRouter;
    this.intervalMs = intervalMs;
  }

  run(signal: AbortSignal): Promise<void> {
    return runPeriodically(this.intervalMs, signal, () => this.worldRouter.runAITickAcrossSectors());
  }
}
